Non uniform fog with perlin-type noise.

lhofhansl opened the pull request:
Something I was playing with.
Just uses a free GLSL perlin noise function and applies it to the fog density.
Self adjusts for the fog range (so the distinct changes of fog density are nice at various view ranges).
Also has a time component so that fog slowly moves (along the z axis, like clouds and waves).

~~Experimental...~~ Have a look. It adds quite a lot of atmosphere, all handled in the shader.

If we like it, I can put it behind an option like the other shader options.

TODO:
~~1. Optional via shader option.~~ Done
~~2. Match cloud movement (which can be configured)...?~~ Decided not to do.
~~3. Special handling under water and in caves...?~~ Done.

lhofhansl commented:
Should probably be not applied underground and underwater.

DS-Minetest commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -210,6 +288,10 @@ void main(void)
 	// Note: clarity = (1 - fogginess)
+	float noise = cnoise(vec3(worldPosition.x * 5 / fogDistance + animationTimer * 20.0,
+							  worldPosition.y * 5 / fogDistance,
+							  worldPosition.z * 5 / fogDistance));
```
Tab, then space alignment is bad for different tab sizes. I suggest to just indent once.

DS-Minetest commented:
Tested.
Dungeons feel so ghostly:
![screenshot_20190902_132029](https://user-images.githubusercontent.com/7613443/64111416-eef63200-cd84-11e9-9bdb-b03920f73634.png)
The landscape looks a bit patchy:
![screenshot_20190902_132207](https://user-images.githubusercontent.com/7613443/64111422-f0bff580-cd84-11e9-920b-c788a5d03668.png)
That one can only see the 3D fog where it traverses node surfaces makes this ugly.
And it's weird that the fog is moving perpendicular to the clouds.

lhofhansl commented:
Thanks for taking a look. That's exactly the feedback I'm looking for.
The patchiness can be changed. I will say that I like it bit patchy, that way it's visible.
Also I will make sure perlin fog is not used below the water line (i.e. under water and in dungeons).
I'll look into the cloud movement.

I should also say that this is a simple, lightweight thing. The fog is not modeled as a thing in the world, there's no ray-tracing to determine the fog level at each point, etc. This just changes the density at every point, which is as close to the "real thing" as one can with such a simple thing.

Calinou commented:
I like the idea, but I think the effect needs to be more subtle as @DS-Minetest said.

lhofhansl commented:
Do you guys want play with this and see what "patchiness" looks nice?
```
float noise = cnoise(vec3(worldPosition.x * 5 / fogDistance + animationTimer * 20.0,
				  worldPosition.y * 5 / fogDistance,
				  worldPosition.z * 5 / fogDistance));
```
The 5 is the "number of patches", if you decrease it you get fewer and bigger patches.
You can also change the fog_start in the settings. It defaults to 0.4, if you decrease it it is smoothed out a bit more.

Perhaps try fog_start = 0.2 in the settings, and change the 5 to 3 and play around with it.

I'll also do some tweaking.

(It turns out that on the client I do not know the water line. The server would have to send it. Perhaps it's easier to reduce the effect when the light is low.)

Lastly... I do not feel emotionally attached to this :) Just an experiment to see what's possible.

lhofhansl commented:
Made the patchiness constant.

With PR, viewrange 240:
![screenshot_20190902_095408](https://user-images.githubusercontent.com/446130/64127738-d7f51700-cd67-11e9-933c-16fb199436bf.png)

Without PR, viewrange 240:
![screenshot_20190902_095441](https://user-images.githubusercontent.com/446130/64127739-d7f51700-cd67-11e9-8f0b-15df170d797a.png)

With PR, viewrange 100:
![screenshot_20190902_095517](https://user-images.githubusercontent.com/446130/64127742-d88dad80-cd67-11e9-99cd-602b797ced7b.png)

Without PR, viewrange 100:
![screenshot_20190902_095446](https://user-images.githubusercontent.com/446130/64127741-d7f51700-cd67-11e9-8cfe-2366974d8250.png)

Calinou commented:
The patchiness looks quite disturbing with high view ranges. It can make taking screenshots of large buildings much harder if you don't want fog to be in the view.

lhofhansl commented:
> To solve this, the patchiness should only be present in the area that would be fogged if the shader is disabled.

Not sure that's possible. The whole point of this is to have uneven fog. :)
I could go back and adjust the patchiness based on the fog range (as the original change had), though I'd think that'd be weird in other scenarios. I think in the end this should just be configurable as the other shader stuff.

(BTW. now that that I am thinking about this, the perlin noise could also be applied to waves and plant movement (removing that weird thing we have now). But that's for another PR)

paramat commented:
This is good that you are experimenting, but i don't think this makes sense, as this isn't actually 'fog', even through it may be called 'fog' in the code. It is an atmospheric haze due to view range distance, which is therefore uniform, and should be uniform to fade out the visible world just before the cutoff of world rendering. The haze gives the landscape a sense of distance, so varying it with noise leads to a distorted sense of distance.

I think any 'fog' feature has to be completely independent of view range haziness. 'Fog' is ideally implemented as low cloud, so should have cuboid volumes and be rendered the same way. The suitable implementation is to have a 2nd level of clouds, and controlled by the cloud API.

Any fog feature would be a fundamental feature so cannot be done by shaders, because many do not or cannot use shaders. For example, it obviously would not be right for clouds to depend on shaders.
Sorry, i don't like being critical like this, and i really appreciate that you are experienting.

lhofhansl commented:
I closed this PR... This will just lead to endless discussion and then not go anywhere. :)

Personally I think this *adds* depth to an otherwise perfectly sterile haze, the haze will still line up perfectly with the horizon and the fog still increases with distance as before (both because the density is adjusted, not the range), and haze is not uniform in nature either.

All good.

lhofhansl commented:
For posterity... The fog in motion.

![vokoscreen-2019-09-02_14-52-14](https://user-images.githubusercontent.com/446130/64134249-22d75480-cd91-11e9-9cef-fe2f5c31e107.gif)

lhofhansl commented:
Ahh shucks... Lemme leave it open so that others see it and comment. I do not plan to do anything else on this, though.

Regardless of this issue, I think MT is woefully falling behind the state of the art. Any change to the graphics is greeted with "Hmm. Nah, won't work on old machines", or "Hmm. Nah, we first need to fix XYZ.", or "Hmm. Nah, it looks different than it looked before.", "Hmm. But what if XYZ?!", etc, etc.

As stated I'm not specifically attached to this one, but generally to the ability and willingness to improve MT visually without perfectionism.

rubenwardy commented:
I like this, but I think it puts fog too close. Maybe a less extreme effect would be better

I would like MT to improve visually, but it would be nice to fix the broken use of OpenGL first

DevyHeavy commented:
I'm curious, is there any way there this could be built into Minetest's LUA api so that modders can choose how it looks? Keep the normal uniform fog that obscures the unloaded world but also let modders have this as an option.

I think it looks amazing. One thing that bothers me about Minetest is not being able to easily modify the 'atmosphere' of the game. Right now, all modders really have is clouds color/density and sky color (you can't even change the sun and the moon yet).

lhofhansl commented:
I think that would be great. Interacting with something as deep "down" as shaders in the engine from Lua is hard, though.

(Sometime like) This can easily be enabled or disabled in the settings, though. Just like waves, wavy plants, bump mapping, etc.
Do you think this is a modding thing, or something each player/client should enable/disable?

TumeniNodes commented:
I love this! Is there a performance hit?
The gif image would give a better visual ref if waving leaves was off, it throws it off due to the effect. Makes it look like the fog is throwing a weird "x-ray" glitch (which it is not but some may think it is when viewing)

*Off topic but, The fire in the background has me greatly intrigued, mind sharing somewhere (even via email?)

lhofhansl commented:
Had to check on the fire. (This is a random test world with all kinds of random old and new mods in it.)
I think what you see is a creature from the doom mod (I think) that leaves behind a particle trail:
![fire](https://user-images.githubusercontent.com/446130/64192634-fde5ee80-ce2f-11e9-9b35-f6177b2f601d.png)

Regarding this. I got the math slightly wrong. Perlin varies from -1 to +1, so it can "move" the haze too close when it's negative. I'll play around with this as I find time.

lhofhansl commented:
I'm having fun with this... Here's what it would like if this was strictly less fog.
With PR (view range 240, fogstart 0.4):
![screenshot_20190903_123816](https://user-images.githubusercontent.com/446130/64203308-faf6f800-ce47-11e9-940a-6ac2f6c077d7.png)

Without PR:
![screenshot_20190903_123932](https://user-images.githubusercontent.com/446130/64203311-fb8f8e80-ce47-11e9-8ff8-05e464c8487d.png)

View range 100, with PR:
![screenshot_20190903_123847](https://user-images.githubusercontent.com/446130/64203309-faf6f800-ce47-11e9-8246-ce7b9328f8aa.png)

And without
![screenshot_20190903_123914](https://user-images.githubusercontent.com/446130/64203310-fb8f8e80-ce47-11e9-97e9-3a233161aace.png)

It's much more subtle this way.

lhofhansl commented:
And here with MT's fogStart parameter set to 0 (view range 240)
![screenshot_20190903_124522](https://user-images.githubusercontent.com/446130/64203591-a902a200-ce48-11e9-83be-13eede4dcfa4.png)
![screenshot_20190903_124557](https://user-images.githubusercontent.com/446130/64203592-a902a200-ce48-11e9-9b78-ec13df4863c5.png)

lhofhansl commented:
I found a simpler (and much faster) noise function. At higher resolutions I could see a perf defect before. I also made the effect much more subtle. Have a look.

With PR, view range 100:
![screenshot_20190905_122245](https://user-images.githubusercontent.com/446130/64374336-79829f80-cfd8-11e9-97fc-2f6a298848d3.png)

Without PR:
![screenshot_20190905_122356](https://user-images.githubusercontent.com/446130/64374339-7a1b3600-cfd8-11e9-8ac8-31b299381b9e.png)

With PR, view range 240:
![screenshot_20190905_122300](https://user-images.githubusercontent.com/446130/64374337-79829f80-cfd8-11e9-983d-4dfab9429632.png)

Without PR:
![screenshot_20190905_122342](https://user-images.githubusercontent.com/446130/64374338-79829f80-cfd8-11e9-844e-8606e8916a2e.png)

I think if I add this as a shader option - just like waving plants - this should be good.
It's not changing the game play or anything, it's not adding new clouds or fog, it just enhances the mood a bit - just like waving plants or smooth lighting.

@paramat I agree the old effect looked too much like (fake) clouds. The effect now just makes the haze a little less regular, and hence it doesn't look so sterile anymore.

@rubenwardy What needs to be fixed in our use of OpenGL? Perhaps I'll find some time to look at that, too.

All, rereading my comment above it sounded more whiny than intended. :)

lhofhansl commented:
And made it configurable. Default off.

lhofhansl commented:
Added options to configure the effect. Also avoid the effect when it's dark (deep under water and in caves).

I do not plan to do anything more on this (until I get any more feedback).
My son and I will have this patch applied and default on :)

v-rob commented:
Honestly, this looks very good.  I would love some more GIFs, since I can only compile on a virtual machine and therefore get ~10 fps.  This is really is a much better atmospheric feeling than the current fog, which feels more like just a view range limiter.

DS-Minetest commented:
Tested again (only with default settings, I'm lazy).
With default settings (and non uniform fog on, of course) the effect is quite subtle, but looks somewhat nice.

There's a problem though.
Sometimes, also in older versions of this PR, I had the feeling that the moving fog jumps back.
Looking into code, I can see that the animation timer is reset every 100 seconds:
https://github.com/minetest/minetest/blob/0b4f424f414380b7a46270e1389e8aa0524d8fba/src/client/game.cpp#L496-L499

lhofhansl commented:
@DS-Minetest I noticed that too and came to the same conclusion. (The waves also jump sometimes.)
I haven't thought of a way around this, yet.

benrob0329 commented:
Reverse the direction of the animation rather than resetting it (eased of course)? It wouldn't look great for waves (probably, if it's eased out a lot it might be fine) but it's definitely better than jumping back to a hard reset.

lhofhansl commented:
Actually the % operation is not even needed. 
```
-               u32 animation_timer = porting::getTimeMs() % 100000;
+               u32 animation_timer = porting::getTimeMs();
		float animation_timer_f = (float)animation_timer / 100000.f;
```
It's still a float, we just need a number changing with time. Continues to work with waves (just sine waves) and perlin noise as both have periodicity build in anyway.

We should just make this change.

paramat commented:
> Do you think this is a modding thing, or something each player/client should enable/disable?

I think it's best as it is, in shaders.
> It's not changing the game play or anything, it's not adding new clouds or fog, it just enhances the mood a bit - just like waving plants or smooth lighting.

Yes i agree.
> I agree the old effect looked too much like (fake) clouds. The effect now just makes the haze a little less regular

It does look much better now, being subtle and larger scale.

Since this is not, and cannot be considered biome fog (it can't be for many reasons i can explain if necessary) i would prefer it is not called 'fog' in the settings, maybe 'haze' or 'mist' or something? To not confuse with biome-fog that could come later, or can be done already using ground-level clouds, as in the Hidden Worlds game.

Does this still drift in a certain direction, or does it just vary with time in a non-directional way? The cloud API can alter the wind direction so a slow drift in a certain direction may look weird.

DS-Minetest commented:
The problem is that, without the modulo, the animation timer might become less precise. (`porting::getTimeMs()` returns a `u64`.)

lhofhansl commented:
Thanks @paramat. Agreed it's not biome fog, happy to call it haze (but then we already have fog_start, which was also added by me... Sorry :( Happy to change that too)

It's still moving in one direction (positive Z). I could add an API for that (perhaps overkill) or config options, or perhaps just take the cloud param values.

How does one alter the cloud direction? I find no configs for that. Can mods do that, only?

It would be nice if this would be just a simple mood-enhancing feature, like waves (which also do not follow any wind direction, then again it's different there). Perhaps it just shouldn't move? (the effect is much reduced when it's stationary, though).

lhofhansl commented:
@DS-Minetest Perhaps. A few moments ago we were at 1568061530474, so 15680615.30474 divided by 100.000. We're not interested in the size of the value, but it's change, the .30474 part, the precision of the mantissa is not changed from before, the larger value just moves into the exponent. So I'm fairly sure it's OK. In fact it works fine with the current value.
I can subtract 1568060000000 or something from the count to be safe, but I do not think it's needed.

paramat commented:
'haze' is fine for me, or maybe any other better word you can think of.

Yes i should quickly add, i'm not requesting the use of the cloud API wind direction. I also somewhat feel it's not worth doing, water waves and waving flora also use a fixed direction. Since it's subtly-changing ground haze it's probably not a big issue to differ from wind direction.

So i have no objection to the current movement. Just one request, the default cloud speed is 2 nodes per second, so i think the haze speed should be the same or lower, not faster.

Also to clarify, i don't officially disapprove of this PR, i'm neutral.

TumeniNodes commented:
"mist"

DS-Minetest commented:
>the larger value just moves into the exponent

What exactly do you mean with this?

>the precision of the mantissa is not changed from before

The bigger the number is, the more of the mantissa is used for the big values. There are less bits available for precise small changes.
If a client would be on a server for very very long time (or even if the computer that the client runs on is already on for a long time), the movement might be juddery or jumpy.

>the default cloud speed is 2 nodes per second, so i think the haze speed should be the same or lower, not faster.

This is the speed that the haze, fog, mist or whatever moves with: `animationTimer * 2 * BS` (Btw. does the `BS` really make sense here?)
`animationTimer` is in 100 seconds (see https://github.com/minetest/minetest/pull/8889#issuecomment-529498322), it increases with `1/100s`. Hence the speed is `20/100=0.2` nodes per second.

lhofhansl commented:
@DS-Minetest Yeah, BS makes no sense there. I had 20 there at some point before instead of 2 * BS.

Re: float, I just meant to say that the pure size of the number is of no significance, and that we just need something that changes. But you're right, we'd run into precision problems.

Another option is to increase the period from 100,000 to 1,000,000 (reset every 16 minutes) or even 10,000,000 (~ every 2h47min, and this is well within float precision). The divider (speed) would remain 100,000.

lhofhansl commented:
@paramat It's OK if you disapprove. I apologize for my reaction above.

Your math is correct @DS-Minetest , however, it looks like the haze is moving than the clouds.
It also depends on the periodicity of the noise function. Note that I'm dividing the world position to make the "blobs" a pleasing size. The divider comes to about 200 or so; in that case the fog would move at about 4n/s. I can half that to match the cloud speed.

Fixer-007 commented:
Actually this fog could be the beginning of weather in vanilla minetest, it does not need to be persistent all day, but show up in some situations, like early morning, evening, at rain, etc

lhofhansl commented:
What do you folks think should happen at night? Looks like there about 5% daylight at night, which means the night_ration < 0.99 will be false it will generate dark haze patches at night. It's not incorrect.

@Fixer-007 It's just a mood-enhancer. I agree with @paramat that proper weather needs to be modeled as actual objects (rain, snow, fog objects, different clouds, etc). Of course the non-uniform haze could be integrated later.

lhofhansl commented:
On the naming of "fog" it's all over minetest already. F3 turns "fog" on an off, etc.
Perhaps if we wanted to change that it should be a different PR with (presumably) some amount of discussion.

I think this one is ready. It's optional, default off, and not that much code.

paramat commented:
Fixer-007
> Actually this fog could be the beginning of weather in vanilla minetest,

It can't be, as sort of explained earlier, as it is clientside and therefore world, biome and server-independent. Weather has to be server controlled and biome-dependent.
> On the naming of "fog" it's all over minetest already. F3 turns "fog" on an off, etc.

Yes i started to realise that too, so 'fog'' is ok, at least for now.

lhofhansl commented:
> Looks like there about 5% daylight at night

My statement doesn't make sense, of course. The night ratio is a feature of the position in the generated map and just indicated how little daylight *can* reach this place (not how much actually does).

ClobberXD commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -14,13 +15,43 @@ varying vec3 eyeVec;
+
+// Simple, fast noise function.
+// See: https://gist.github.com/patriciogonzalezvivo/670c22f3966e662d2f83
+vec4 perm(vec4 x){return mod(((x * 34.0) + 1.0) * x, 289.0);}
```
```suggestion
vec4 perm(vec4 x) { return mod(((x * 34.0) + 1.0) * x, 289.0); }
```

DS-Minetest replied:
All that copied code has horrible code style.

lhofhansl replied:
Fixed

lhofhansl commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -257,7 +257,9 @@ local function formspec(tabview, name, tabdata)
 			"checkbox[8.25,3.5;cb_waving_plants;" .. fgettext("Waving Plants") .. ";"
-					.. dump(core.settings:get_bool("enable_waving_plants")) .. "]"
+					.. dump(core.settings:get_bool("enable_waving_plants")) .. "]" ..
+			"checkbox[8.25,4.0;cb_fog_noise;" .. fgettext("Non Uniform Fog") .. ";"
```
Done

lhofhansl commented:
@DS-Minetest The change is good as is, I think... Just for completeness I did more testing on floats. It looks like floats are accurate on whole number to exactly 2^24 or 16.777.216. Since we're dividing by 100.000. We have about 1677721600000 or about 53 years of precision.

lhofhansl commented:
Also. Does MT reset the worldPosition, i.e. the GL coordinates of the world? It certainly looks like. If I move around every now and then the fog jumps a bit. I ruled out the animation time for this. So it somehow must reset the world/model coordinates.

How do I get the actual node coordinates?

I think this is still, but it can be improved.

paramat commented:
> Does MT reset the worldPosition, i.e. the GL coordinates of the world?

Yes, every 200 nodes or so, it's a fix for visual bugs occuring when far from world centre.

lhofhansl commented:
Thanks @paramat . Hmm, so how does get actual world coordinates in the shader? I could pass in the offset, I suppose, but that would defeat the purpose.

paramat commented:
Not sure how.
I've also noticed jumps in shader behaviour, i assumed it was the animation timer resetting.

lhofhansl commented:
Is that also costly. Perhaps it explains some of the stuttering I see when moving around a lot.

Perhaps make this configurable? 1000 seems a better default anyway. Yes, some minor glitches might be visible at the extremes.

paramat commented:
As far as i know it isn't costly.
I also suggested 1000 years ago, but it was claimed that 200 was quite well chosen and apparently the rendering glitches occur at surprisingly small distances from 0,0,0. Haven't tested that though.

Stuttering when moving around is much more likely to be the load on mesh creation, that is what causes FPS stutter for me.

lhofhansl commented:
I see. Hmmm...

Any opposition to releasing this one as is? Or is it the "jumping" of the fog an unacceptable glitch?

lhofhansl commented:
OK... I have a change, which passes the current camera offset to the shaders. That way the shaders can decide what they want to do... If they want to use the actual coordinate, just add the passed cameraOffset to worldPosition. That's generally good to have!

For the fog the lack of precision works fine, because (a) it's fuzzy anyway, and (b) it's not doing any further transformation on it other using it for the noise function. I tested this at 20000 positions, and it all looks fine.

Should I update this PR with that?

lhofhansl commented:
Heck. Did it anyway. As I said the shaders might generally want to have a choice on whether to use the adjusted or original coordinates.

paramat commented:
> passes the current camera offset to the shaders. That way the shaders can decide what they want to do... If they want to use the actual coordinate, just add the passed cameraOffset to worldPosition. That's generally good to have!

This is useful, and good to see it wasn't complex to do.
So later we will be able to fix jumps in the other shaders.

lhofhansl commented:
Yep. Currently this is only used for the waves and non-uniform fog.

Good to merge this one?

lhofhansl commented:
Anything else I need to do here?
(Before merge I'd add a comment to the shaders explaining the meaning of worldPosition and cameraOffset.)

paramat commented on src/client/shader.cpp:
```diff
@@ -686,6 +686,16 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
+		shaders_header += itos(g_settings->getU16("fog_noise_frequency") / 10);
+		shaders_header += "\n";
+		shaders_header += "#define FOG_NOISE_STRENGTH ";
+		shaders_header += ftos(g_settings->getFloat("fog_noise_strength") * 4.0);
```
4.0f

paramat commented:
No further requests from me, once my line comment is attended to.

lhofhansl commented:
Thanks @paramat . Fixed the number format, and also added comments to the shaders explaining the use of both worldPosition and cameraOffset for future shader effects.

lhofhansl commented:
(When this is merged I have a proposal PR for Perlin waves that are configured with exactly the same parameters as the current waves, but instead of simple (and sterile) sine waves these use Perlin noise. No more work for the shader and look more realistic.)

But it needs some of the changes here.

lhofhansl commented:
Let's get this in. The change to pass the cameraOffset to the shaders is needed for other issues as well.

If need be I can make a *separate PR* just for that, but since I went through great length to make this configurable with small changes only we should just pull this in.

Let me know...

lhofhansl commented:
Hmm... I always mess things up when I do a rebase :(

lhofhansl commented:
So do we want this? :)
The only controversial behavior could potentially be in semi-dark places that are artificially lit, since I have nothing to go by other than the day-night ratio to decide whether this effect should happen or not.
(But even then it's only when this feature is enabled)

(@paramat , I'd love to think about how mods can control (at least enable/disable) shader effects in a future PR. This would go well with your new weather mod for example (and be totally lightweight on the server). The client would have to opt in of course.)

paramat commented:
I suspect core devs are busy with release, so hopefully this will get more attention after release.
I dislike the concept as explained above, but it is a harmless shader so i am not going to officially disapprove.

I think this shader is incompatible with my MTG weather intentions (wrong wind speed and direction for a start), and i would not want to have anything to do with it because i dislike the concept (sorry). Users can enable the shader if they want but that has to be the choice of the user as some users cannot use shaders or do not like them. So i think this shader being independently switched, as currently, is best, also simplest.

SmallJoker commented:
The fog effect is quite subtle, but might bring a little bit of variety to the game. No concept objections from my side.

lhofhansl commented:
Ping...

lhofhansl commented:
Hi all. Is there still interest in this?

Perhaps the effect is too subtle now? (it's configurable, I'd just have to change some default.
Please let me know. I'm also OK closing this.

paramat commented:
It seems liked by 2 core devs, rubenwardy posted 'i like this' above, and SmallJoker thumbed-up the first post.

hecktest commented:
What if this was also world space height based? That could look good.
Also consider #9895, generic effect APIs are more powerful than hardcoded effects, and the effort is better spent letting anyone make their own cool fog shader.

paramat commented:
The shader eye position bugfix should perhaps be split off from this into a new PR to fix bumpmapping issues. Bugfixes and features should ideally be separate commits anyway:
From https://github.com/minetest/minetest_game/issues/2473#issuecomment-534350169
https://github.com/minetest/minetest_game/issues/2473#issuecomment-534842743

> Edit: In fact it looks like the lightVec is miscalculated. The eyePosition passed to the shader is global position, whereas the worldPosition is relative to the camera offset. In minetest/minetest#8889 I pass the cameraOffset to the shaders so that real position can be calculated. We should either do that, or adjust the eyePosition with the cameraOffset before passing it to the shaders.

> We need the change from minetest/minetest#8889 for me to fix this, or alternatively I can create a engine PR just for passing the cameraOffset.

The delay in reviewing this PR is delaying the bugfix, and it seems the bugfix is fairly simple which means it will get reviewed quicker (i will review and test a bugfix PR, but i cannot approve this feature PR).

paramat commented:
Hmm but ...
See https://github.com/minetest/minetest/issues/9711#issuecomment-706549399
If we decide to improve the shader bumpmapping light source positioning, as needs to be done, that alters the same code, so we should decide on that first.
